### ISSUE TITLE
Configure mini-browser to return to PowerShell 7

### ISSUE THREAD

**Issue by tinojr** [PowerShell/PowerShell#19772]
### Summary of the new feature / enhancement

Hello,
At my organization we have separate admin accounts and windows SSO enabled using 3rd party federation. So, it's not easy to log into admin portals, and very difficult to connect to M365/AZ PowerShell modules with my admin account. Sure, a dedicated admin workstation might help, but security team has delayed that option. Generally, when I need to log into an admin portal, I have to use a private browser to log on. With PowerShell 5, I wrote a function that will open as Administrator and as my admin account, then when I connect to M365/AZ modules the mini-browser already detects my admin account. The behavior is different with PowerShell 7, it opens to my default browser even if I'm running pwsh as my admin account. Due to SSO, it doesn't work because it detects my regular account. I've been using the -UseWindowsPowerShell in PowerShell 7 as a workaround. Although, this doesn't seem to be most ideal since it's enabling lots of compatibility settings and requires PowerShell 5 - all I really need is the mini-browser so it correctly detects the account running pwsh (my admin account). I looked at about_Windows_PowerShell_Compatibility and about_PowerShell_Config but it's not clear (or maybe i missed it) if this is configurable - it doesn't appear to be. So hoping you can add an enhancement.
Thanks,
Tino

### Proposed technical implementation details (optional)

Add a switch to Import-Module that allows us to call a mini-browser (such as PowerShell 5 behavior) that excludes all the other compatibility settings in -UseWindowsPowerShell.  Alternatively, provide a way to modify the behavior in powershell.config.json file.

**Comment by iSazonov** [comment]
Could you please add more simple repro steps and screenshots? Can you demo the issue with starting locally something like Notepad?

**Comment by tinojr** [comment]
Yes, Let me get that info. I have no issues opening local apps such as notepad, it's only with cloud modules that redirect to a browser for authentication. Let me get some screenshots.

**Comment by tinojr** [comment]
Hello, here are the steps to reproduce. I've included a side by side comparison between my admin account (C---) and my regular account (T---) in both PowerShell 5 and PWSH 7. Since this is public forum, I've blocked out some confidential information. If you would like to screen-share for a live session, let me know.

**Implement Seamless SSO with Windows**
We have [Okta](https://www.okta.com/partners/microsoft/windows-10/) integrated with our Windows 10 systems, but I believe you can implement a similar experience with [AAD Seamless SSO](https://learn.microsoft.com/en-us/azure/active-directory/hybrid/connect/how-to-connect-sso). The experience will be a seamless logon experience for the logged-on Windows user when attempting to authenticate into any application that supports this feature. So, while I am logged into Windows 10 as my regular user (non-admin), I cannot log into portal.azure.com or admin.microsoft.com with alternate admin credentials. The work-around for that is to open a private browser and enter admin credentials. This is more difficult to do in PowerShell/PWSH, see below for details. Additionally, PWSH is unable to log on with alternate credentials unless using compatibility feature that proxies to PowerShell (-UseWindowsPowerShell).

**SSO Experience with PowerShell 5**
1.	**Open PowerShell**. I wrote a local function to logon as ‘Administrator’ and my admin account in same PowerShell/pwsh session, but here is the main part of code to get you started. For my regular session, I just opened PowerShell shortcut as ‘Administrator’.
$MsgPrompt = "Enter password for Admin account"
$PS_EXE_Chosen = 'PowerShell.exe' # 'pwsh.exe'
$FullUserName = 'contoso\AdminUser' #domain needed for this
[string]$ArgumentList = '-command "Start-Process ' + $PS_EXE_Chosen + ' -Verb RunAs"'
Start-Process -FilePath $PS_EXE_Chosen -Credential (Get-Credential -UserName $FullUserName -Message $MsgPrompt) -ArgumentList $ArgumentList
 
  
![image](https://github.com/PowerShell/PowerShell/assets/32583391/e484a646-6adc-4a6a-9f77-a8d7f767fde4)


2.	**Connect to M365/Azure in session**. In this example, import ExchangeOnlineManagement module, then connect to EXO. Notice that my regular account session detects I am logged into Windows. Even if I click ‘use another account’ in my regular session and specify my admin account, windows sso does not give me opportunity to enter password because it autologins using my account that is logged into windows (regular account).
 
![image](https://github.com/PowerShell/PowerShell/assets/32583391/9a8d13fd-3029-4e8e-a4bc-a9b426d272e5)


3.	**Observe logon.** In our organization, we have okta, so you will see the okta mfa prompt on my admin session. However, in my regular session it just seamlessly sso’d into EXO without allowing me to specify admin credentials.
 
![image](https://github.com/PowerShell/PowerShell/assets/32583391/c85349cb-f387-4bc9-a301-ef6a229b25df)

4.	**Confirm accounts logged on**. Confirm my admin session is logged in with admin account and confirm regular session did not logon as admin (even though I tried). In my admin session you can see I query another user. In my regular session you can see I my regular account does not have permission to query another user. (Note: In Exchange, default is anyone can log into PowerShell and query their own account.)
 
![image](https://github.com/PowerShell/PowerShell/assets/32583391/61370495-68f7-438e-b6a4-8125b6ccad87)




**SSO Experience with PowerShell 7**
1.	**Open PWSH**. I wrote a local function to logon as ‘Administrator’ and my admin account in same PowerShell/pwsh session, but here is the main part of code to get you started. For my regular session, I just opened PWSH shortcut as ‘Administrator’.
$MsgPrompt = "Enter password for Admin account"
$PS_EXE_Chosen = 'pwsh.exe' # 'PowerShell.exe' 
$FullUserName = 'contoso\AdminUser' #domain needed for this
[string]$ArgumentList = '-command "Start-Process ' + $PS_EXE_Chosen + ' -Verb RunAs"'
Start-Process -FilePath $PS_EXE_Chosen -Credential (Get-Credential -UserName $FullUserName -Message $MsgPrompt) -ArgumentList $ArgumentList

 
![image](https://github.com/PowerShell/PowerShell/assets/32583391/70256f41-9ca1-4f4f-8464-357496dfb3d0)


2.	**Connect to M365/Azure in session**. In this example, import ExchangeOnlineManagement module, then connect to EXO. Here is where I need to use compatibility switch ([UseWindowsPowerShell](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_windows_powershell_compatibility?view=powershell-7.3)) in order to get the mini-browser in my admin session. As I understand, it is proxying a connection to PowerShell, not ideal. Also, notice that my regular account session detects I am logged into Windows and opens the default browser in a large window. Even if I click ‘use another account’ in my regular session and specify my admin account, windows sso does not give me opportunity to enter password because it autologins using my account that is logged into windows (regular account).

 
![image](https://github.com/PowerShell/PowerShell/assets/32583391/0dabc94a-4f00-41f9-bbfd-58ca605c0954)


3.	**Observe logon.** In our organization, we have okta, so you will see the okta mfa prompt on my admin session. However, in my regular session it just seamlessly sso’d into EXO without allowing me to specify admin credentials even after I input my admin account. The windows is left open and informing me to close browser.
 
![image](https://github.com/PowerShell/PowerShell/assets/32583391/926f060c-dec2-4f2b-a7eb-1288fb1a53d1)

4.	**Confirm which accounts are logged on.** Confirm my admin session is logged in with my admin account and confirm regular session did not logon as admin (even though I tried). In my admin session you can see I can query another user. In my regular session you can see I do not have permission to query another user. There is an unrelated error with EXO V3 module in PWSH admin session, you can ignore (just started happening). (Note: In Exchange, default is anyone can log into PowerShell and query their own account.)

![image](https://github.com/PowerShell/PowerShell/assets/32583391/af62c8a9-85f7-427b-9d5d-fc52ffd7b1a3)

**Comment by iSazonov** [comment]
@tinojr Thanks for detailed description!

Have you tried `Connect-ExchangeOnline -UserPrincipalName  <your_admin_account>` fom regular session?

**Comment by tinojr** [comment]
Yes I have, that is overridden by Windows SSO. ;-)

**Comment by iSazonov** [comment]
Yes, It'd be great to get Okta support thoughts about the issue. Or we need a repro without the Okta software.

**Comment by tinojr** [comment]
Ok, it seems like my point is getting missed. This is not unique to okta nor to exchangeonlinemanagement module. Seamless SSO is doing what it is designed to do.  The problem is that the authentication behavior has changed between powershell.exe (5) and pwsh.exe (7), hence the known work-around that is called out here -> (https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_windows_powershell_compatibility?view=powershell-7.3#using-the-compatibility-feature) .  This problem with this work-around is that it proxies all connections back to powershell.exe (5), so it's a little unreliable.  I am hoping someone from powershell development team can add this as a feature request for pwsh - the ability to have the same authentication behavior as powershell.exe (mini-browser ).

**Comment by iSazonov** [comment]
The repro steps contain two problematic points:
1. Proxying is used. In fact, the module does not work in PowerShell 7. This means that we need to solve the problem with EXO dev group.
2. Third-party soft is used. I am sure that any MSFT group will ask to reproduce the problem without this software. That is why I advised to start with Okta support. (It may turn out that they do not work correctly with .Net Core.)

Only research by the Okta dev group and MSFT EXO dev group can reveal the root of the problem. Until then it is not at all clear that the problem is in the pwsh7 code.

**Comment by kilasuit** [comment]
@tinojr - apologies if you felt that we'd missed the ask, that wasn't the case and was trying to help you in finding a suitable workaround in the meantime, other than the one you've listed, as is the case with issue triaging in a repo as busy as this. Especially if the ask requires dev work then I can't say how quickly this would be added to the set of ongoing work.
There is a possibility that this is functionality in EXO module that hasn't been migrated across, and I say that as I believe that the [PnP.PowerShell module](https://github.com/pnp/powershell) (used for SharePoint and some teams stuff) does infact prompt with that mini-browser. I could be mistaken here & if so then yes there would be a need for this to be looked as a feature request, if it is at all possible & it may have been pulled due to changes in dotnet, but I don't think this to be the case.

**Comment by confidentlyincorrec** [comment]
It's a bit perplexing to me that this behavior was changed in Pwsh 7, when there was nothing wrong with using WebView 2 (aka mini-browser) in Powershell 5, perhaps someone could enlighten me? 

@tinojr's situation is not unique (I'd wager most well-run organizations have separate accounts for admin access) and why I found this thread. My default web browser is Edge, and uses my 'daily' account. I use Chrome for my 'admin' account, that way I keep things nicely segregated and don't have to constantly log-in to InPrivate/Incognito windows. 

Equally, this isn't unique to the Exchange Online PowerShell module, it'll affect all other Microsoft (SharePoint, Graph, etc) management modules.

All this to say, an option should be added to PowerShell 7 to set the authentication behaviour to _always use Webview2_, if that is the user's preference.

**Comment by GDogg79** [comment]
I've just run into the exact same issue as described above. Running PS 7.4.1. MS auth flows are pushed to edge browser that I'm authenticated to using SSO (regular account with no elevated privileges). Even if I select "sign in as a different user", SSO overrides this and persists with regular account. My only workaround at the moment is to force edge to launch in InPrivateMode by default. 

Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Policies\Microsoft\Edge
DWORD 32 --> InPrivateModeAvailability ; value = 2

Essentially this means I must use another browser client for regular account. This is highly undesirable behaviour.

Has any progress been made in this space? Does PS 7.5 address this issue?

**Comment by Solid-Duff** [comment]
@tinojr I too am facing the same exact problem you are facing with an almost identical setup. Have you been able to find a solution for this issue?

**Comment by zaaj** [comment]
I've just started trying to use Powershell 7, and have been running into these problems as well.  I use Edge - default profile is my regular user account, I have separate Edge user profiles for admin accounts, and the connect-<service> commands call up my non-privileged-user edge profile window.
I have NOT tried setting Edge to use the most recently opened profile to open hyperlinks, because I NEVER want to open random links with my admin profiles.

I HAVE had success with the -device option to Connect-<service>, since it gives me a device code I can use, but it means I have to:
1) manually open my admin browser profile
2) navigate to https://microsoft.com/devicelogin
3) paste in the device code presented in the Pwsh window
4) confirm that I wanted to log into the service name
5) close the now-useless browser tab

vs in Windows Powershell, using WebView2, the first service I connect to pops up a standalone mini-browser I can use to enter admin credentials, subsequent service connections I just select the already-logged-in admin credential and connect.  The mini browser closes and I'm set to work.

Also, not all services have the -Device option - Connect-MicrosoftTeams is the one I ran into today that does not.

My understanding is that Edge is available cross platform, so is WebView2 also cross-platform?  Could WebView2 be called from the Powershell service admin modules in Pwsh like it is in Windows Powershell?  That would help a lot.

Barring that ideal, having a way to configure which browser / browser profile is used for service connection authorization would appear to be important for admins trying to stick to the principle of least priviledge security best practice.

**Comment by bastienperez** [comment]
This behavior is problematic as it severely impacts security. Many organizations enforce separate admin accounts, and PowerShell 7's authentication model forces administrators to either weaken security or resort to unreliable workarounds. While dedicated admin workstations might be a theoretical solution, in practice, few small or medium-sized businesses can afford them.

The previous PowerShell 5 behavior, which allowed WebView2 for authentication, was much more flexible and aligned with best practices. Removing it without providing an alternative is a step backward.

**Comment by confidentlyincorrec** [comment]
Well said, @bastienperez. This seems like trivially easy change for the maintainers to make. The bare minimum should be to give users the option to _always use WebView2_

Someone please fork this and revert the change, I cry out in vain to the universe.
